---
title: "Comparing E-G predictions with CRISPR data"
date: "`r format(Sys.time(), '%B %d, %Y')`"
params:
  rmd: "comparePredictionsToExperiment.Rmd"
output:
  html_document:
    number_sections: yes
    toc: yes
---

```{r setupDocument, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
# save.image("comp.rda")
# stop()
```

```{r attachPackages}
# attach required packages and functions
library(tidyverse)
library(cowplot)
library(DT)
library(ROCR)
library(caTools)
library(UpSetR)
source(file.path(snakemake@scriptdir, "crisprComparisonLoadInputData.R"))
source(file.path(snakemake@scriptdir, "crisprComparisonPlotFunctions.R"))
source(file.path(snakemake@scriptdir, "crisprComparisonBootstrapFunctions.R"))
```

```{r prepareInputData}
# load merged data
merged <- fread(snakemake@input$merged, colClasses = c("ValidConnection" = "character"))

# import pred_config file
include_col <- ifelse(is.null(snakemake@params$include_col), "include", snakemake@params$include_col)
pred_config <- importPredConfig(snakemake@input$pred_config,
                                expr = "baseline.nearestExprTSS" %in% merged$pred_uid,
                                include_col = include_col)

# process merged data for benchmarking analyses, including filtering for ValidConnection == TRUE
merged <- processMergedData(merged, pred_config = pred_config, filter_valid_connections = TRUE,
                            include_missing_predictions = snakemake@params$include_missing_predictions,
                            distToTSS_as_kb = TRUE)

# only retain pred_config entries also in merged data (important when not using an include column)
pred_config <- subset(pred_config, pred_uid %in% unique(merged$pred_uid))
check_unique_identifier(pred_config, col = "pred_name_long")

# extract colors for predictors in all plots
if (all(is.na(pred_config$color))) {
  pred_colors <- NULL
} else {
  pred_colors <- deframe(select(pred_config, pred_name_long, color))
}

# column identifying experimental positives
pos_col <- snakemake@params$pos_col

# directory to save plots in addition to adding them to report
plotdir <- file.path(dirname(snakemake@output[[1]]), "plots")
```

This is the output for the comparison **`r snakemake@wildcards$comparison`**. Following analyses
evaluate how well the experimental data agrees with the predictions of CRE - gene pairs. Following
input files were used:

```{r}
# extract predictions and experiment input files from config in snakemake object
comp_config <- snakemake@config$comparisons[[snakemake@wildcards$comparison]]
expt_file <- basename(comp_config$expt)
pred_file <- basename(unlist(comp_config$pred))
```

Experimental data: **`r expt_file`**  
Predictions: **`r pred_file`**

Following parameters in the config file **`r snakemake@input$pred_config`** were used to overlap
predictions with experimental data and to assess performance of predictors. If no config file was
provided, this was generated using default values. It's strongly recommended to use a prediction
config file to control how predictors should be treated.
```{r printPredConfig}
options(htmlwidgets.TOJSON_ARGS = list(na = 'string'))
pred_config_print <- pred_config[, -c("include", "pred_uid")]
datatable(pred_config_print, options = list(pageLength = 20), autoHideNavigation = TRUE)
```

***

# Overlap between predictors and CRISPR data {.tabset .tabset-pills}
The number of CRISPR enhancer-gene pairs that overlapped enhancer-gene pairs for each predictor are
counted. CRISPR enhancer-gene pairs that did not overlap any predicted pairs, are considered not
predicted. Large fractions of CRISPR E-G pairs not overlapping predictions lead to poor performance.
```{r overlaps, fig.height=5, fig.width=7}
# count and plot number of CRISPR E-G pairs overlapping prediction E-G pairs
overlap_plots <- applyCellTypes(merged, .fun = plotOverlaps)

# save plots to files
savePlotList(overlap_plots, basename = "expt_pred_overlaps.pdf", path = plotdir, height = 3.5,
             width = 7)
```

```{r results='asis', fig.cap=cap, fig.height=3.5, fig.width=7}
# print plots for every cell type in tabs
printTabbedPlots(overlap_plots, section_level = "#")

# figure caption
cap <- paste("Number of CRISPR enhancer-gene pairs overlapping enhancer-gene pairs in predictions.")
```

***

# Precision-Recall performance
Precision-recall (PR) curves are used for comparing the performance of different predictors on the
experimental data. The area under the PR curve (AUPRC) provides a single metric of a predictors
performance.

```{r computePRC}
# compute precision-recall tables for all cell types
pr <- applyCellTypes(merged, .fun = calcPRCurves, pred_config = pred_config, pos_col = pos_col)

# combine pr tables into one table and save to file for other downstream analyses
pr_table <- rbindlist(pr, idcol = "ExperimentCellType")
write_tsv(pr_table, file = file.path(dirname(snakemake@output[[1]]), "pr_table.txt.gz"))

# create performance summary tables
perf_summary <- applyCellTypes(merged, .fun = makePRSummaryTableBS, pred_config = pred_config,
                               pos_col = pos_col)

# convert performance summaries to one table and add full predictor names to performance summary
perf_summary <- perf_summary %>% 
  bind_rows(.id = "cell_type") %>% 
  left_join(select(pred_config, pred_uid, pred_name_long), by = "pred_uid") %>% 
  relocate(pred_name_long, .after = pred_uid)

# save performance summary to text file
write_tsv(perf_summary,
          file = file.path(dirname(snakemake@output[[1]]), "performance_summary.txt"))
```

```{r plotPRC}
# calculate number and percentage of experimental true positives in the experimental dataset
n_pos <- applyCellTypes(merged, .fun = calcNPos, pos_col = pos_col)
pct_pos <- applyCellTypes(merged, .fun = calcPctPos, pos_col = pos_col)

# get number of tested enhancer gene pairs in merged data and create title for PR plot
n_pairs <- applyCellTypes(merged, .fun = function(df) n_distinct(df$name) )
pr_title <- paste0(snakemake@wildcards$comparison, " (", unlist(n_pairs[names(pr)]), " pairs)")

# make PRC plots
n_pos <- n_pos[names(pr)]
pct_pos <- pct_pos[names(pr)]
pr_plots <- mapply(FUN = makePRCurvePlot, pr_df = pr, n_pos = n_pos, pct_pos = pct_pos, plot_name = pr_title,
                   MoreArgs = list(pred_config = pred_config,
                                   min_sensitivity = snakemake@params$min_sensitivity,
                                   line_width = 0.8, point_size = 3,
                                   text_size = 13, colors = pred_colors),
                   SIMPLIFY = FALSE)

# save plots to files
savePlotList(pr_plots, basename = "prc_full_expt_data.pdf", path = plotdir, height = 4.5,
             width = 8)
```

## Precision-recall curves {.tabset .tabset-pills}

```{r results='asis', fig.cap=cap, fig.height=4.5, fig.width=8}
# print plots for every cell type in tabs
printTabbedPlots(pr_plots, section_level = "##")

# figure caption
cap <- paste("Precision-recall curves for all predictors in all matching experimental cell types.",
             "Dots represent alpha cutoff values as specified in pred_config file. If no alpha was",
             "set, the minium alpha in predictions was taken by default, respectively the maximum",
             "for inverse predictors. Distance to TSS was added as baseline predictor and computed",
             "from the provided 'gene universe'.")
```

## Performance summary

```{r perfSummary, fig.cap=cap}
# pretty-print PR summary table for report
perf_summary  %>% 
  select(-pred_uid) %>%
  rename(predictor = pred_name_long) %>% 
  mutate(across(where(is.double), round, digits = 3)) %>% 
  datatable(
    perf_summary,
    extensions = c("FixedColumns", "Buttons"),
    options = list(
      pageLength = 20,
      dom = 'Bfrtip',
      scrollX = TRUE,
      fixedColumns = list(leftColumns = 3),
      buttons = c('copy', 'csv', 'excel', 'pdf', 'print')
    ),
    autoHideNavigation = TRUE
    ) %>%
  formatStyle(c(2, 5, 9), `border-right` = "solid 1px")

# figure caption
cap <- paste("Precision-recall performance summary for predictors. Table shows Area-under-the-PRC",
             "(AUPRC) and precision at specified thresholds (if specified) and minimum sensitity",
             "(recall) of 0.7.")
```

***

# Receiver Operating Characteristic performance
Receiver Operating Characteristic (ROC) curves are an alternative method to compare performance by
computing true positive rates and false positive rates for each predictor.

```{r plotROC}
# make ROC curve plots for each cell type
roc_plots <- applyCellTypes(merged, .fun = plotROC, pos_col = "Regulated",
                            pred_config = pred_config, line_width = 0.8, point_size = 3,
                            text_size = 13, colors = pred_colors)

# save plots to files
savePlotList(roc_plots, basename = "roc_full_expt_data.pdf", path = plotdir, height = 4.5,
             width = 8)
```

```{r results='asis', fig.cap=cap, fig.height=4.5, fig.width=8}
# print plots for every cell type in tabs
printTabbedPlots(roc_plots, section_level = "##")

# figure caption
cap <- paste("ROC curves for all predictors in all matching experimental cell types.",
             "Distance to TSS and nearest genes/TSS were added as baseline predictors and computed",
             "from the provided 'gene universe'.")
```

***

# Effect size vs predictors {.tabset .tabset-pills}
Each predictor listed in the prediction data is plotted against the effect size of enhancer
perturbations reported in the experimental data (e.g. percent change in expression). These plots
show how well a predictor is associated with effects observed in CRISPRi enhancer screens in an
intuitive way.

```{r effectSizeScatter, include=FALSE}
# make predictor vs effect size scatter plots
es_scatters <- applyCellTypes(merged, .fun = plotPredictorsVsEffectSize, pos_col = pos_col,
                              pred_names_col = "pred_name_long", point_size = 2, text_size = 13,
                              alpha_value = 1, label.x.npc = 0.7, ncol = 2)

# calculate plot dimensions based on rows and column in plot
es_scatter_dims <- get_row_col(es_scatters[[1]])
plot_height <- ceiling(es_scatter_dims[1] * 2.5)
plot_width  <- ceiling(es_scatter_dims[2] * 3.5)

# save scatter plots to files
savePlotList(es_scatters, basename = "EffectSize_scatter_plots.pdf", path = plotdir,
             height = plot_height, width = plot_width)
```

```{r, results='asis', fig.cap=cap, fig.height=plot_height, fig.width=plot_width}
# print plots for every cell type in tabs
printTabbedPlots(es_scatters, section_level = "#")

# figure caption
cap <- paste("Predictors versus CRISPRi effect size. Effect size is defined as percent change in",
             "target gene expression upon CRISPRi perturbation of an enhancer. Effect size values",
             "are taken from 'EffectSize' column in experimental data, while predictor scores",
             "correspond to scores from prediction files. Numbers show Spearman's rank correlation",
             "coefficient (rho) between effect size and predictor scores.")
```

***

# Predictor scores versus experimental outcome {.tabset .tabset-pills}
The scores of each predictor is compared between experimental positives and negatives to get another
assessment of how well it distinguishes true enhancer - gene pairs from negatives.

```{r}
# create plots showing predictor values as a function of experimental outcome
pred_plots <- applyCellTypes(merged, .fun = plotPredictorsVsExperiment, pos_col = pos_col,
                             pred_names_col = "pred_name_long", text_size = 13)

# calculate plot dimensions based on number of predictors
pred_plots_dims <- get_row_col(pred_plots[[1]])
plot_height <- ceiling(pred_plots_dims[1] * 2.5)
plot_width  <- ceiling(pred_plots_dims[2] * 3.5)

# save scatter plots to files
savePlotList(pred_plots, basename = "predictor_vs_experiment.pdf", path = plotdir,
             height = plot_height, width = plot_width)
```

```{r, results='asis', fig.cap=cap, fig.height=plot_height, fig.width=plot_width}
# print plots for every cell type in tabs
printTabbedPlots(pred_plots, section_level = "#")

# figure caption
cap <- paste("Predictor scores vs. experimental outcome for all predictors. Each point represents",
  "one E-G pair in the experimental data. Cases where the predictor value is 0 or infinite might",
  "correspond to E-G pairs that were not found in predictions and predicor values were filled in",
  "according to the prediction config file")
```

***

# Performance as function of distance to TSS
Enhancer-gene pairs are binned based on their distance to TSS and predictor performance is assessed
for each bin.

```{r distanceBins, fig.width=4.5, fig.height=3}
# set distance bins to 4 equal sized bins if no manual boundaries are provided
if (is.null(snakemake@params$dist_bins_kb)) {
  dist_breaks <- 4
} else {
  dist_breaks <- as.numeric(snakemake@params$dist_bins_kb)
}

# bin pairs by distance
dist_bins <- merged %>% 
  filter(pred_uid == "baseline.distToTSS") %>% 
  mutate(`distToTSS (bins)` = cut(pred_value, breaks = dist_breaks, right = FALSE)) %>% 
  select(name, `distToTSS (bins)`)

# add distance bins to merged
merged_bins <- merged %>% 
  left_join(dist_bins, by = "name") %>% 
  filter(!is.na(`distToTSS (bins)`))

# get unique experimentally tested pairs (TODO: implement count_pairs_subset() for this)
crispr_pairs <- distinct(select(merged_bins, name, `distToTSS (bins)`, all_of(pos_col)))

# plot number of pairs per distance bin
ggplot(crispr_pairs, aes(x = `distToTSS (bins)`, fill = get(pos_col))) +
  geom_bar() +
  labs(x = "Distance to TSS (kb)", y = "Number of E-G pairs",
       title = "CRISPR E-G pairs vs. distance", fill = pos_col) +
  scale_fill_manual(values = c("FALSE" = "darkgray", "TRUE" = "steelblue")) +
  theme_bw()

# save plot to file
ggsave(filename = file.path(plotdir, "eg_pairs_vs_distance.pdf"), width = 4.5, height = 3)
```

```{r filterBins}
# count CRISPR positives and negatives per bin
crispr_pos_neg_pairs <- crispr_pairs %>%
  group_by(`distToTSS (bins)`) %>%
  summarize(pos = sum(get(pos_col) == TRUE),
            neg = sum(get(pos_col) == FALSE))

# get any bins containing no positives or no negatives
invalid_bins <- crispr_pos_neg_pairs %>% 
  filter(pos == 0 | neg == 0) %>% 
  pull(`distToTSS (bins)`)

# remove these from merged data, since performance can't be assessed for these
merged_bins_filt <- merged_bins %>% 
  filter(!`distToTSS (bins)` %in% invalid_bins) %>% 
  mutate(`distToTSS (bins)` = droplevels(`distToTSS (bins)`))

# report if any bins are removed
if (length(invalid_bins) > 0) {
  message("Following distance bins are removed because they do not contain any CRISPR positives ",
          "or negatives: ", paste(invalid_bins, collapse = ", "))
}
```

## AUPRC as function of distance {.tabset .tabset-pills}
```{r AUPRCVsDist, warning=FALSE}
# compute performance as a function of distance to TSS
dist_perf <- applyCellTypes(merged_bins_filt, .fun = computePerfSubsets, pred_config = pred_config,
                            subset_col = "distToTSS (bins)", metric = "auprc", pos_col = pos_col,
                            bs_iter = 1000)

# create table of performance across distance
dist_perf_tbl <- bind_rows(dist_perf, .id = "cell_type")
write_tsv(dist_perf_tbl,
          file = file.path(dirname(snakemake@output[[1]]), "performance_summary_distance_binned.txt"))

# plot performance as a function of distance to TSS
dist_auprc <- lapply(dist_perf, FUN = plotPerfSubsets, pred_config = pred_config,
                     subset_name = "Distance to TSS (kb)", title = "AUPRC vs. distance to TSS")

# save distance prc plots to files
savePlotList(dist_auprc, basename = "distance_auprc.pdf", path = plotdir, height = 5, width = 8)
```

```{r, results='asis', fig.cap=cap, fig.height=5, fig.width=8}
# print plots for every cell type in tabs
printTabbedPlots(dist_auprc, section_level = "##")

# figure caption
cap <- paste("Area under the Precision-Recall Curve (AUPRC) for different distance to TSS bins (kb).")
```

## Precision-recall curves {.tabset .tabset-pills}
```{r PRCVsDist}
# count the number of distance bins to set number of rows in plot grid
n_bins <- n_distinct(merged_bins_filt$`distToTSS (bins)`)
nrow <- ceiling(n_bins / 2)

# create PR curves per distance bin for all cell types
dist_prc_plots <- applyCellTypes(merged_bins_filt, .fun = makePRCurveSubsets,
                                 subset_cols = "distToTSS (bins)", pred_config = pred_config,
                                 pos_col = pos_col,
                                 min_sensitivity = snakemake@params$min_sensitivity,
                                 line_width = 0.8, point_size = 3, text_size = 13, nrow = nrow,
                                 colors = pred_colors)

# calculate plot dimensions based on number of features
plot_height <- nrow * 4
plot_width <- ceiling(n_bins / nrow) * 7

# save distance prc plots to files
savePlotList(dist_prc_plots, basename = "distance_binned_prc.pdf", path = plotdir,
             height = plot_height, width = plot_width)
```

```{r, results='asis', fig.cap=cap, fig.height=plot_height, fig.width=plot_width}
# print plots for every cell type in tabs
printTabbedPlots(dist_prc_plots, section_level = "##")

# figure caption
cap <- paste("Precision-Recall curves for different distance to TSS bins (kb).")
```

## Predictor scores versus experimental outcome {.tabset .tabset-pills}
```{r PredVsExpVsDist}
# create plots showing predictor values as a function of experimental outcome
dist_pred_plots <- applyCellTypes(merged_bins_filt, .fun = plotPredVsExperimentSubsets,
                                  subset_cols = "distToTSS (bins)", pos_col = pos_col,
                                  pred_names_col = "pred_name_long", text_size = 13)

# calculate plot dimensions based on number of features and predictors
plot_width <- ceiling(n_bins * 2.5)
plot_height <- ceiling(n_distinct(merged_bins_filt$pred_name_long) * 1.5)

# save distance predictor vs. experiment plots to files
savePlotList(dist_pred_plots, basename = "distance_binned_pred_vs_expt.pdf", path = plotdir,
             height = plot_height, width = plot_width)
```

```{r, results='asis', fig.cap=cap, fig.height=plot_height, fig.width=plot_width}
# print plots for every cell type in tabs
printTabbedPlots(dist_pred_plots, section_level = "##")

# figure caption
cap <- paste("Predictor scores versus experimental outcome for different distance to TSS bins",
             "(kb).")
```

## Effect size vs predictors {.tabset .tabset-pills}
```{r EsVsPredVsDist}
# create plots showing effect size as function of predictor scores
dist_es_plots <- applyCellTypes(merged_bins_filt, .fun = predVsEffectSizeSubsets,
                                subset_cols = "distToTSS (bins)", pos_col = pos_col,
                                pred_names_col = "pred_name_long", point_size = 2, text_size = 16,
                                label.x.npc = 0.65)

# calculate plot dimensions based on number of features and predictors
plot_height <- ceiling(n_bins * 2.5)
plot_width <- ceiling(n_distinct(merged_bins_filt$pred_name_long) * 2)

# save distance effect size vs. predictor plots to files
savePlotList(dist_es_plots, basename = "distance_binned_EffectSize_scatter.pdf", path = plotdir,
             height = plot_height, width = plot_width)
```

```{r, results='asis', fig.cap=cap, fig.height=plot_height, fig.width=plot_width}
# print plots for every cell type in tabs
printTabbedPlots(dist_es_plots, section_level = "##")

# figure caption
cap <- paste("CRISPR effect size vs predictor scores for different distance to TSS bins (kb).")
```

***

# Subset by gene and enhancer features
If any gene or enhancer features are provided versions faceted by these features of the PR curves,
predictor vs experiment and effect size plots are created.

```{r featureCols}
# columns in merged data containing gene and enhancer features
gene_feat_cols <- grep(colnames(merged), pattern = "^gene_feature.+$", value = TRUE)
enh_feat_cols <- grep(colnames(merged), pattern = "^enh_feature_.+$", value = TRUE)

# plot the number of positive and negative pairs per gene and enhancer feature
if (length(c(gene_feat_cols, enh_feat_cols)) > 0) {
  
    # count the number of positive and negative pairs per gene and enhancer feature
    n_pairs_features <- countPairsFeatures(merged, pos_col = pos_col)
    
    # plot the number of pairs per feature type
    n_pairs_plots <- plotPairsFeatures(n_pairs_features)
    
}

# calculate dimensions of plot based on number of rows and columns in faceted plots
if (length(gene_feat_cols) > 0) {
  gene_plot_dims <- get_row_col(n_pairs_plots$Gene)
  gene_plot_height <- ceiling(gene_plot_dims[1] * 2.5)
  gene_plot_width  <- ceiling(gene_plot_dims[2] * 2.5)
} else {
  gene_plot_height <- 5
  gene_plot_width  <- 7
}

if (length(enh_feat_cols) > 0) {
  enh_plot_dims <- get_row_col(n_pairs_plots$Enhancer)
  enh_plot_height <- ceiling(enh_plot_dims[1] * 2.5)
  enh_plot_width  <- ceiling(enh_plot_dims[2] * 2.5)
} else {
  enh_plot_height <- 5
  enh_plot_width  <- 7
}
```

```{r, fig.height=gene_plot_height, fig.width=gene_plot_width}
if (length(gene_feat_cols) > 0) {
  n_pairs_plots$Gene
}
```

```{r, fig.height=enh_plot_height, fig.width=enh_plot_width}
if (length(enh_feat_cols) > 0) {
  n_pairs_plots$Enhancer
}
```

## Precision-recall curves

### Gene features {.tabset .tabset-pills}
```{r geneFeaturePRC}
if (length(gene_feat_cols) > 0) {
 
  # create subset plots for all cell types
  gene_feat_plots <- applyCellTypes(merged, .fun = makePRCurveSubsets, subset_cols = gene_feat_cols,
                                    pred_config = pred_config, pos_col = pos_col,
                                    min_sensitivity = snakemake@params$min_sensitivity,
                                    line_width = 1.2, point_size = 3.5, text_size = 13,
                                    colors = pred_colors)
  
  # calculate plot dimensions based on number of features
  n_features <- length(gene_feat_cols)
  plot_height <- ceiling(n_features * 3.5)

}
```

```{r, results='asis', fig.cap=cap, fig.height=plot_height, fig.width=10}
if (length(gene_feat_cols) > 0) {
  
  # print plots for every cell type in tabs
  printTabbedPlots(gene_feat_plots, section_level = "###")
  
  # figure caption
  cap <- paste("Precision-Recall curves for subsets of the data based on provided gene features.")
  
}
```

### Enhancer features {.tabset .tabset-pills}
```{r enhFeaturePRC}
if (length(enh_feat_cols) > 0) {
  
  # create subset plots for all cell types
  enh_feat_plots <- applyCellTypes(merged, .fun = makePRCurveSubsets, subset_cols = enh_feat_cols,
                                   pred_config = pred_config, pos_col = pos_col,
                                   min_sensitivity = snakemake@params$min_sensitivity,
                                   line_width = 1.2, point_size = 3.5, text_size = 13,
                                   colors = pred_colors)
  
  # calculate plot dimensions based on number of features
  n_features <- length(enh_feat_cols)
  plot_height <- ceiling(n_features * 3.5)

}
```

```{r, results='asis', fig.cap=cap, fig.height=plot_height, fig.width=10}
if (length(enh_feat_cols) > 0) {

  # print plots for every cell type in tabs
  printTabbedPlots(enh_feat_plots, section_level = "###")
  
  cap <- paste("Precision-Recall curves for subsets of the data based on provided enhancer",
  "features.")

}
```

## Predictor scores versus experimental outcome

### Gene features {.tabset .tabset-pills}
```{r geneFeaturePredVsExp}
if (length(gene_feat_cols) > 0) {

  # create plots showing predictor values as a function of experimental outcome
  pred_plots <- applyCellTypes(merged, .fun = plotPredVsExperimentSubsets, 
                               subset_cols = gene_feat_cols, pos_col = pos_col,
                               pred_names_col = "pred_name_long", text_size = 13)
  
  # calculate plot dimensions based on number of predictors
  plot_height <- ceiling(n_distinct(merged$pred_name_long) * length(gene_feat_cols) * 1.5)

}
```

```{r, results='asis', fig.cap=cap, fig.height=plot_height, fig.width=7}
if (length(gene_feat_cols) > 0) {
  
  # print plots for every cell type in tabs
  printTabbedPlots(pred_plots, section_level = "###")
  
  cap <- paste("Predictor scores vs. experimental outcome for all predictors for subsets of the",
               "data based on provided gene features.")
  
}
```

### Enhancer features {.tabset .tabset-pills}
```{r enhFeaturePredVsExp}
if (length(enh_feat_cols) > 0) {
  
  # create plots showing predictor values as a function of experimental outcome
  pred_plots <- applyCellTypes(merged, .fun = plotPredVsExperimentSubsets,
                               subset_cols = enh_feat_cols, pos_col = pos_col,
                               pred_names_col = "pred_name_long", text_size = 13)
  
  # calculate plot dimensions based on number of predictors
  plot_height <- ceiling(n_distinct(merged$pred_name_long) * length(enh_feat_cols) * 1.5)
  plot_height <- min(c(plot_height, 170))
  
}
```

```{r, results='asis', fig.cap=cap, fig.height=plot_height, fig.width=7}
if (length(enh_feat_cols) > 0) {
  
  # print plots for every cell type in tabs
  printTabbedPlots(pred_plots, section_level = "###")
  
  # figure caption
  cap <- paste("Predictor scores vs. experimental outcome for all predictors for subsets of the data",
               "based on provided gene features")

}
```

## Effect size vs predictors

### Gene features {.tabset .tabset-pills}
```{r geneFeatureEsScatter}
if (length(gene_feat_cols) > 0) {
  
  # create plots showing effect size as function of predictor scores
  es_plots <- applyCellTypes(merged, .fun = predVsEffectSizeSubsets, subset_cols = gene_feat_cols,
                             pos_col = pos_col, pred_names_col = "pred_name_long", point_size = 2,
                             text_size = 16, label.x.npc = 0.65)
  
  # calculate plot dimensions based on number of predictors
  plot_height <- ceiling(length(gene_feat_cols) * 5.75)
  plot_width <- ceiling(n_distinct(merged$pred_name_long) * 2)

}
```

```{r, results='asis', fig.cap=cap, fig.height=plot_height, fig.width=plot_width}
if (length(gene_feat_cols) > 0) {

  # print plots for every cell type in tabs
  printTabbedPlots(es_plots, section_level = "###")
  
  # figure caption
  cap <- paste("Effect size vs. predictor scores for all predictors for subsets of the data based on",
               "provided gene features")

}
```

### Enhancer features {.tabset .tabset-pills}
```{r enhFeatureEsScatter}
if (length(enh_feat_cols) > 0) {
  
  # create plots showing effect size as function of predictor scores
  es_plots <- applyCellTypes(merged, .fun = predVsEffectSizeSubsets, subset_cols = enh_feat_cols,
                             pos_col = pos_col, pred_names_col = "pred_name_long", point_size = 2,
                             text_size = 16, label.x.npc = 0.65)
  
  # calculate plot dimensions based on number of predictors
  plot_height <- ceiling(length(enh_feat_cols) * 5.75)
  plot_width <- ceiling(n_distinct(merged$pred_name_long) * 2)

}
```

```{r, results='asis', fig.cap=cap, fig.height=plot_height, fig.width=plot_width}
if (length(enh_feat_cols) > 0) {

  # print plots for every cell type in tabs
  printTabbedPlots(es_plots, section_level = "###")
  
  # figure caption
  cap <- paste("Effect size vs. predictor scores for all predictors for subsets of the data based on",
               "provided enhancer features")
  
}
```

***

# Correlation between predictors
How well predictor scores correlate with each other for E-G pairs in the experimental data is
investigated.
```{r predCorMatrix}
# create correlation matrix plots for all cell types
corr_matrix_plots <- applyCellTypes(merged, .fun = plotPredCorMatrix,
                                    pred_names_col = "pred_name_long", method = "spearman")

# plot dimensions based on number of predictors
n_pred <- n_distinct(pull(filter(merged, boolean == FALSE), pred_uid))
dim <- 4 + (n_pred - 1) / 3

# save correlation matrix plots to pdfs
savePlotList(corr_matrix_plots, basename = "predictor_correlation.pdf", path = plotdir,
             height = dim, width = dim)
```

```{r, results='asis', fig.cap=cap, fig.height=dim, fig.width=dim}
# print plots for every cell type in tabs
printTabbedPlots(corr_matrix_plots, section_level = "#")

# figure caption
cap <- paste("Correlation of scores between predictors for experimenta E-G pairs.")
```

***

# Properties of the experimental dataset
Different features of the experimental data are investigated.

## Distance to TSS distribution {.tabset .tabset-pills}
```{r distToTSSDistr}
# create distance to TSS distributions for all cell types
dist_distr <- applyCellTypes(merged, .fun = plotDistanceDistribution,
                             dist = "baseline.distToTSS", pos_col = pos_col, text_size = 13)

# save plots to pdfs
savePlotList(dist_distr, basename = "distToTSS_distribution.pdf", path = plotdir, height = 5,
             width = 7)
```

```{r, results='asis', fig.cap=cap, fig.height=5, fig.width=7}
# print plots for every cell type in tabs
printTabbedPlots(dist_distr, section_level = "##")
  
# figure caption
cap <- paste("Distance to TSS distributions for all E-G pairs in experimental data. E-G pairs are",
  "partitioned according to whether they were identified as enhancer-gene interactions (positives)",
  "or negatives.")
```

## Overlapping features {.tabset .tabset-pills}
A plot showing the number of experimentally tested candidate enhancers overlapping provided genomic
features. If no features were provided, this plot is not generated.

```{r enhFeatures}
# columns in merged data containing information on overlapping enhancer features
feature_cols <- grep(colnames(merged), pattern = "^enh_feature_.+$", value = TRUE)

# create upset plots if merged data contains any enhancer feature columns
if (length(feature_cols > 0)) {
  
  # make upset plots from features overlapping experimentally tested enhancers for each cell type
  overlap_plots <- applyCellTypes(merged, .fun = plotOverlappingFeatures,
                                  feature_cols = feature_cols)
  
  # save plots to pdfs
  for (i in names(overlap_plots)) {
    pdf(file.path(plotdir,  paste0(i, ".overlappingFeatures.pdf")), height = 5, width = 7)
    print(overlap_plots[[i]])
    dev.off()
  }
}
```

```{r, results='asis', fig.height=5, fig.width=7, fig.cap=cap}
if (length(feature_cols > 0)) {
  
  # print plots for every cell type in tabs
  printTabbedPlots(overlap_plots, section_level = "##", plot_function = print)

  # figure caption
  cap <- paste("Genomic features overlapping experimentally tested enhancers.")
  
}
```

***

# Sources
* <a download="comparePredictionsToExperiment.Rmd" href="`r base64enc::dataURI(file = params$rmd, mime = 'text/rmd',
    encoding = 'base64')`">R Markdown source file (to produce this document)</a>
